CHANGELOG.md

```diff
@@ -22,6 +22,8 @@ The following emojis are used to highlight certain changes:
 
 ### Fixed
 
+* `mfs`: directory cache is now cleared every time the directory node is read, somewhat limiting unbounded growth and time to sync it to the underlying unixfs.
```

Comment:
I'm scared to replace known MFS problems with new unknown MFS problems, but also not familiar with codebase well enough to have informed opinion :(

I know folks have pipelines that do a lot of concurrent writes to MFS via ipfs files CLI/RPC and likely figured out what works and what does not.

Are we sure new version removes or at least decreases surface for data loss during concurrent writes?

Perhaps useful context: in kubo, we have --flush flag that (if true) calls mfs.FlushPath after each MFS operation.

Intention behind disabling flush is to allow user to avoid sync during many small operations and flush to disk until "final" version of file or directory is ready, but that comes at consistency risk if there is more than one thread doing mutation:

Most of the subcommands of 'ipfs files' accept the '--flush' flag. It defaults
to true. Use caution when setting this flag to false. It will improve
performance for large numbers of file operations, but it does so at the cost
of consistency guarantees.

So effectively, patching a lot of things should be faster with flush=false but you need to do things sequentially. This is the mental model Kubo users have.

Are edge cases mentioned in  https://github.com/ipfs/boxo/pull/751/files avoidable  with  --flush=true ? Or are we introducing racy risk there, where it was not before?

If we make "safe" default less "safe", could/should we tie cache cleanup to happen only in  mfs.FlushPath instead (explicit --flush=true)? (call syncCache with true only if in the context of explicit mfs.FlushPath?)

Reply:
That might be an option... 

I did not know we flushed by default... that also means this commit will essentially kill the cache on every operation. I cannot say with certainty but this probably brings the new "unknown MFS problems" as it will be equivalent to disabling the cache.

Let's do something, let me spend another while trying to figure out why exactly concurrent writes to a single file break with things like cache-size=1 and if something else is broken that way. Maybe I can find a workaround to feel better about it.